### PR TITLE
fix: Catch ECONNRESET and other errors more reliably

### LINKED ISSUE #1146: http.py should catch more ECONNRESET conditions
Right now http.py implicitly assumes that an `ECONNRESET` error will be raised as a `ConnectionError`.  However, in some cases `urllib3` will raise this a regular OSError with the ernno set to match `ECONNRESET`.  http.py should check for the errnos that correspond to the subclasses of ConnectionError to be extra safe.

### PR DESCRIPTION
**Pull request by dossett**
Fixes #1146  🦕

### CODE REVIEW

**Comment by dossett** [comment]
@googlebot I signed it!

**Comment by monoueka** [comment]
Hi

**Comment by monoueka** [comment]
SreyOn

File: googleapiclient/http.py [hunk 1]
```diff
@@ -190,11 +187,16 @@ def _retry_request(
...
             # errno's contents differ by platform, so we have to match by name.
+            # Some of these same errors may have been caught above, e.g. ECONNRESET *should* be
+            # raised as a ConnectionError, but some libraries will raise it as a socket.error
+            # with an errno corresponding to ECONNRESET
             if socket.errno.errorcode.get(socket_error.errno) not in {
                 "WSAETIMEDOUT",
                 "ETIMEDOUT",
                 "EPIPE",
                 "ECONNABORTED",
+                "ECONNREFUSED",
```
Comment (parthea) [inline googleapiclient/http.py:198]: @dossett Thanks for the fix! The test [TestHttpRequest.test_no_retry_connection_errors](https://github.com/googleapis/google-api-python-client/blob/master/tests/test_http.py#L937) is failing. I think it's because the test is [expecting OSError to be raised ](https://github.com/googleapis/google-api-python-client/blob/master/tests/test_http.py#L174) when the `socket.errno` is `ECONNREFUSED`. The test passes if I remove `ECONNREFUSED` from this list. Please can you take a look?

File: tests/test_http.py [hunk 2]
```diff
@@ -174,10 +174,10 @@ def request(self, *args, **kwargs):
             # set errno to a non-retriable value
             try:
                 # For Windows:
-                ex.errno = socket.errno.WSAECONNREFUSED
+                ex.errno = socket.errno.WSAEHOSTUNREACH
```
Comment (dossett) [inline tests/test_http.py:177]: Since `ECONNREFUSED` is now retried, we need to switch these to a different error that won't be retried

File: googleapiclient/http.py (hunk 1 above)
Comment (dossett) [reply googleapiclient/http.py:198]: Thanks for pointing that out @parthea.  I made a small change to `test_http` to have it use `EHOSTUNREACH` for errors that shouldn't be retried.  That's just the error that caught my eye when I looked through a list of error codes, but it does seem to work ok.

**Comment by dossett** [comment]
I'll squash all the commits to conform to the commit message expectations when/if this is ok to merge.

File: .gitignore [hunk 3]
```diff
@@ -6,6 +6,7 @@ dist/
 
 # Test files
 .nox/
+venv/
```
Comment (parthea) [inline .gitignore:9]: `venv` is likely the name of your specific virtual environment. Can we revert this change?

**Review by parthea — COMMENTED** [review]
Thanks @dossett ! Just one minor observation.

File: .gitignore (hunk 3 above)
Comment (dossett) [reply .gitignore]: whoops, thanks for spotting that!

**Comment by monoueka** [comment]
Srey

**Review by parthea — COMMENTED** [review]
Hi @dossett , The changes look good. Please could you update the commit message to have prefix `fix:`?

**Comment by dossett** [comment]
@parthea commit message updated, thanks!

**Review by parthea — APPROVED** [review]
LGTM, Thanks @dossett !
